File: python/sagens/_decode.py
from __future__ import annotations

from pathlib import Path
from uuid import UUID

from ._models import (
    AdminCredentialBundle,
    BoxBooleanSetting,
    BoxCredentialBundle,
    BoxNumericSetting,
    BoxRecord,
    BoxRuntimeUsage,
    BoxSettings,
    BoxStatus,
    CheckpointRestoreMode,
    ExecExit,
    FileKind,
    FileNode,
    ManagedDaemonPaths,
    ManagedDaemonStartInfo,
    ReadFileResult,
    UserConfig,
    WorkspaceChange,
    WorkspaceChangeKind,
    WorkspaceCheckpointRecord,
    WorkspaceCheckpointSummary,
)


def user_config_from_dict(raw: dict) -> UserConfig:
    return UserConfig(
        version=raw["version"],
        admin_uuid=UUID(raw["admin_uuid"]),
        admin_token=raw["admin_token"],
        endpoint=raw["endpoint"],
    )


def admin_bundle_from_dict(raw: dict) -> AdminCredentialBundle:
    return AdminCredentialBundle(
        admin_uuid=UUID(raw["admin_uuid"]),
        admin_token=raw["admin_token"],
        endpoint=raw["endpoint"],
    )


def box_bundle_from_dict(raw: dict) -> BoxCredentialBundle:
    return BoxCredentialBundle(
        box_id=UUID(raw["box_id"]),
        box_token=raw["box_token"],
        endpoint=raw["endpoint"],
    )


def box_record_from_dict(raw: dict) -> BoxRecord:
    sandbox_id = raw.get("active_sandbox_id")
    return BoxRecord(
        box_id=UUID(raw["box_id"]),
        name=raw.get("name"),
        status=BoxStatus(raw["status"]),
        settings=box_settings_from_dict(raw["settings"]) if raw.get("settings") else None,
        runtime_usage=(
            runtime_usage_from_dict(raw["runtime_usage"])
            if raw.get("runtime_usage")
            else None
        ),
        workspace_path=Path(raw["workspace_path"]),
        active_sandbox_id=UUID(sandbox_id) if sandbox_id else None,
        created_at_ms=raw["created_at_ms"],
        last_start_at_ms=raw.get("last_start_at_ms"),
        last_stop_at_ms=raw.get("last_stop_at_ms"),
        last_error=raw.get("last_error"),
    )


def box_settings_from_dict(raw: dict) -> BoxSettings:
    return BoxSettings(
        cpu_cores=numeric_setting_from_dict(raw["cpu_cores"]),
        memory_mb=numeric_setting_from_dict(raw["memory_mb"]),
        fs_size_mib=numeric_setting_from_dict(raw["fs_size_mib"]),
        max_processes=numeric_setting_from_dict(raw["max_processes"]),
        network_enabled=boolean_setting_from_dict(raw["network_enabled"]),
    )


def runtime_usage_from_dict(raw: dict) -> BoxRuntimeUsage:
    return BoxRuntimeUsage(
        cpu_millicores=raw["cpu_millicores"],
        memory_used_mib=raw["memory_used_mib"],
        fs_used_mib=raw["fs_used_mib"],
        process_count=raw["process_count"],
    )


def numeric_setting_from_dict(raw: dict) -> BoxNumericSetting:
    return BoxNumericSetting(current=raw["current"], max=raw["max"])


def boolean_setting_from_dict(raw: dict) -> BoxBooleanSetting:
    return BoxBooleanSetting(current=raw["current"], max=raw["max"])


def exec_exit_from_raw(raw: str | dict) -> ExecExit:
    if isinstance(raw, str):
        return ExecExit(kind=raw)
    if "exit_code" in raw:
        return ExecExit(kind="exit_code", code=raw["exit_code"])
    kind, value = next(iter(raw.items()))
    return ExecExit(kind=kind, code=value if isinstance(value, int) else None)


def file_node_from_dict(raw: dict) -> FileNode:
    return FileNode(
        path=raw["path"],
        kind=FileKind(raw["kind"]),
        size=raw["size"],
        digest=raw.get("digest"),
        target=raw.get("target"),
    )


def read_file_from_dict(raw: dict) -> ReadFileResult:
    return ReadFileResult(
        path=raw["path"],
        data=bytes(raw["data"]),
        truncated=raw["truncated"],
    )


def workspace_change_from_dict(raw: dict) -> WorkspaceChange:
    kind_after = raw.get("kind_after")
    return WorkspaceChange(
        path=raw["path"],
        kind=WorkspaceChangeKind(raw["kind"]),
        kind_after=FileKind(kind_after) if kind_after else None,
    )


def checkpoint_record_from_dict(raw: dict) -> WorkspaceCheckpointRecord:
    return WorkspaceCheckpointRecord(
        summary=checkpoint_summary_from_dict(raw["summary"]),
        source_checkpoint_id=raw.get("source_checkpoint_id"),
        changes=[workspace_change_from_dict(item) for item in raw["changes"]],
    )


def checkpoint_summary_from_dict(raw: dict) -> WorkspaceCheckpointSummary:
    return WorkspaceCheckpointSummary(
        checkpoint_id=raw["checkpoint_id"],
        workspace_id=raw["workspace_id"],
        name=raw.get("name"),
        metadata=dict(raw.get("metadata", {})),
        created_at_ms=raw["created_at_ms"],
    )


def daemon_start_info_from_dict(raw: dict) -> ManagedDaemonStartInfo:
    return ManagedDaemonStartInfo(
        paths=ManagedDaemonPaths(
            state_dir=Path(raw["paths"]["state_dir"]),
            user_config_path=Path(raw["paths"]["user_config_path"]),
            endpoint=raw["paths"]["endpoint"],
            pid_path=Path(raw["paths"]["pid_path"]),
        ),
        user_config=user_config_from_dict(raw["user_config"]),
        already_running=raw["already_running"],
    )


def serialize_box_setting(name: str, value: bool | int) -> dict:
    setting_name = snake_case(name)
    return {"setting": setting_name, "value": value}


def serialize_restore_mode(mode: CheckpointRestoreMode | str) -> str:
    return mode.value if isinstance(mode, CheckpointRestoreMode) else str(mode)


def snake_case(name: str) -> str:
    return name.strip().replace("-", "_")

File: node/src/decode.ts
import {
  AdminCredentialBundle,
  BoxCredentialBundle,
  BoxRecord,
  BoxSettings,
  BoxRuntimeUsage,
  CompletedExecution,
  ExecExit,
  FileNode,
  ReadFileResult,
  UserConfig,
  WorkspaceCheckpointRecord
} from "./models.js";

export function userConfigFromWire(raw: Record<string, unknown>): UserConfig {
  return {
    version: numberValue(raw.version),
    adminUuid: stringValue(raw.admin_uuid),
    adminToken: stringValue(raw.admin_token),
    endpoint: stringValue(raw.endpoint)
  };
}

export function adminBundleFromWire(raw: Record<string, unknown>): AdminCredentialBundle {
  return {
    adminUuid: stringValue(raw.admin_uuid),
    adminToken: stringValue(raw.admin_token),
    endpoint: stringValue(raw.endpoint)
  };
}

export function boxBundleFromWire(raw: Record<string, unknown>): BoxCredentialBundle {
  return {
    boxId: stringValue(raw.box_id),
    boxToken: stringValue(raw.box_token),
    endpoint: stringValue(raw.endpoint)
  };
}

export function boxRecordFromWire(raw: Record<string, unknown>): BoxRecord {
  return {
    boxId: stringValue(raw.box_id),
    name: optionalString(raw.name),
    status: stringValue(raw.status) as BoxRecord["status"],
    settings: raw.settings ? boxSettingsFromWire(objectValue(raw.settings)) : null,
    runtimeUsage: raw.runtime_usage
      ? runtimeUsageFromWire(objectValue(raw.runtime_usage))
      : null,
    workspacePath: stringValue(raw.workspace_path),
    activeSandboxId: optionalString(raw.active_sandbox_id),
    createdAtMs: numberValue(raw.created_at_ms),
    lastStartAtMs: optionalNumber(raw.last_start_at_ms),
    lastStopAtMs: optionalNumber(raw.last_stop_at_ms),
    lastError: optionalString(raw.last_error)
  };
}

export function execExitFromWire(raw: unknown): ExecExit {
  if (typeof raw === "string") {
    return { kind: raw, success: raw === "success" };
  }
  const obj = objectValue(raw);
  if (typeof obj.exit_code === "number") {
    return { kind: "exit_code", code: obj.exit_code, success: obj.exit_code === 0 };
  }
  const [kind, value] = Object.entries(obj)[0] ?? ["killed", null];
  const code = typeof value === "number" ? value : null;
  return { kind, code, success: kind === "success" || code === 0 };
}

export function completedExecution(
  exitStatus: ExecExit,
  output: Buffer,
  stdout: Buffer,
  stderr: Buffer
): CompletedExecution {
  const exitCode =
    exitStatus.kind === "success" ? 0 : exitStatus.kind === "exit_code" ? exitStatus.code : null;
  return {
    exitStatus,
    exitCode,
    output,
    stdout,
    stderr,
    outputText: output.toString(),
    stdoutText: stdout.toString(),
    stderrText: stderr.toString()
  };
}

export function fileNodeFromWire(raw: Record<string, unknown>): FileNode {
  return {
    path: stringValue(raw.path),
    kind: stringValue(raw.kind) as FileNode["kind"],
    size: numberValue(raw.size),
    digest: optionalString(raw.digest),
    target: optionalString(raw.target)
  };
}

export function readFileFromWire(raw: Record<string, unknown>): ReadFileResult {
  return {
    path: stringValue(raw.path),
    data: Buffer.from(arrayValue(raw.data) as number[]),
    truncated: booleanValue(raw.truncated)
  };
}

export function checkpointFromWire(raw: Record<string, unknown>): WorkspaceCheckpointRecord {
  const summary = objectValue(raw.summary);
  return {
    summary: {
      checkpointId: stringValue(summary.checkpoint_id),
      workspaceId: stringValue(summary.workspace_id),
      name: optionalString(summary.name),
      metadata: recordString(summary.metadata),
      createdAtMs: numberValue(summary.created_at_ms)
    },
    sourceCheckpointId: optionalString(raw.source_checkpoint_id),
    changes: arrayValue(raw.changes).map((item) => {
      const change = objectValue(item);
      return {
        path: stringValue(change.path),
        kind: stringValue(change.kind) as WorkspaceCheckpointRecord["changes"][number]["kind"],
        kindAfter: optionalString(change.kind_after) as
          | WorkspaceCheckpointRecord["changes"][number]["kindAfter"]
          | undefined
      };
    })
  };
}

export function objectValue(value: unknown): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError("expected object value");
  }
  return value as Record<string, unknown>;
}

function boxSettingsFromWire(raw: Record<string, unknown>): BoxSettings {
  return {
    cpuCores: settingNumber(raw.cpu_cores),
    memoryMb: settingNumber(raw.memory_mb),
    fsSizeMib: settingNumber(raw.fs_size_mib),
    maxProcesses: settingNumber(raw.max_processes),
    networkEnabled: settingBoolean(raw.network_enabled)
  };
}

function runtimeUsageFromWire(raw: Record<string, unknown>): BoxRuntimeUsage {
  return {
    cpuMillicores: numberValue(raw.cpu_millicores),
    memoryUsedMib: numberValue(raw.memory_used_mib),
    fsUsedMib: numberValue(raw.fs_used_mib),
    processCount: numberValue(raw.process_count)
  };
}

function settingNumber(value: unknown) {
  const raw = objectValue(value);
  return { current: numberValue(raw.current), max: numberValue(raw.max) };
}

function settingBoolean(value: unknown) {
  const raw = objectValue(value);
  return { current: booleanValue(raw.current), max: booleanValue(raw.max) };
}

function recordString(value: unknown): Record<string, string> {
  if (!value) {
    return {};
  }
  const raw = objectValue(value);
  return Object.fromEntries(Object.entries(raw).map(([key, item]) => [key, stringValue(item)]));
}

function arrayValue(value: unknown): unknown[] {
  if (!Array.isArray(value)) {
    throw new TypeError("expected array value");
  }
  return value;
}

function booleanValue(value: unknown): boolean {
  if (typeof value !== "boolean") {
    throw new TypeError("expected boolean value");
  }
  return value;
}

function numberValue(value: unknown): number {
  if (typeof value !== "number") {
    throw new TypeError("expected number value");
  }
  return value;
}

function stringValue(value: unknown): string {
  if (typeof value !== "string") {
    throw new TypeError("expected string value");
  }
  return value;
}

function optionalNumber(value: unknown): number | null {
  return value === undefined || value === null ? null : numberValue(value);
}

function optionalString(value: unknown): string | null {
  return value === undefined || value === null ? null : stringValue(value);
}
